proc/proc.go

```diff
@@ -0,0 +1,249 @@
+	}
+	cmd := exec.Command(cmdPath, args...)
+
+	var stdout *os.File
+	if stdout, err = ioutil.TempFile("", ""); err != nil {
+		return Proc{}, err
+	}
+	cmd.Stdout = stdout
+	stdoutName := stdout.Name()
+
+	var stderr *os.File
+	if stderr, err = ioutil.TempFile("", ""); err != nil {
+		return Proc{}, err
+	}
+	cmd.Stderr = stderr
+	stderrName := stderr.Name()
+
+	if err = cmd.Start(); err != nil {
+		return Proc{}, err
+	}
+
+	go func() {
+		// Wait on the cmd to make sure resources get released
+		cmd.Wait()
+	}()
+
+	return Proc{
+		stdoutFileName: stdoutName,
+		stderrFileName: stderrName,
+		cmd:            cmd,
+	}, nil
+}
+
+// Kill causes the running process to exit and closes the related resources.
+func (p Proc) Kill() error {
+	if err := p.cmd.Process.Kill(); err != nil {
+		// processes which have ended return errors
+		return err
+	}
+	// Throw away the error from Wait() because an error is returned if either:
+	// 1. Wait() was already called (cool, we just use it to know that the process exited)
+	// 2. The process returned a non-zero exit code (cool, we will return any exit code)
+	p.cmd.Wait()
+	return nil
+}
+
+// Status returns the status of the process.
+func (p Proc) Status() ProcStatus {
+	if p.cmd.ProcessState == nil {
+		return Running
+	}
+
+	if p.cmd.ProcessState.ExitCode() != 0 {
+		return Stopped
+	}
+	return Exited
+}
+
+// ProcStatus is the status of a process.
+type ProcStatus int
+
+const (
+	// Running indicates that the process is running.
+	Running ProcStatus = iota
+	// Exited indicates that the process returned a non-zero exit code.
+	Exited
+	// Stopped indicates that the process returned no exit code.
+	Stopped
+)
+
+func (ps ProcStatus) String() string {
+	return [...]string{"Running", "Exited", "Stopped"}[ps]
+}
+
+// Query streams output from the process to the returned channel.
+// The Stdout and Stderr files are opened for reads and polled until
+// a third routine finds that the process has exited.
+// The third routine cancels the context of the pollReads.
+// After the read routines finish the third routine sends the ExitCode and closes the channel.
+func (p Proc) Query() (<-chan ProcOutput, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	stream := make(chan ProcOutput)
+
+	wg := &sync.WaitGroup{}
+
+	wg.Add(2)
+	if err := pollRead(ctx, p.stdoutFileName, wg, stream, newProcOutput_Stdout); err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to setup poll read on stdout: %w", err)
+	}
+	if err := pollRead(ctx, p.stderrFileName, wg, stream, newProcOutput_Stderr); err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to setup poll read on stderr: %w", err)
+	}
+
+	go func() {
+		// Throw away the error from Wait() because an error is returned if either:
+		// 1. Wait() was already called (cool, we just use it to know that the process completed)
+		// 2. The process returned a non-zero exit code (cool, we will return any exit code)
+		p.cmd.Wait()
+		cancel()
```

Comment:
`cancel` will not cause the reading to stop in pollRead, which is confusing, because that's what we expect when we cancel a context. I think this might have been introduced to handle cases where we're reading the output of a completed process, but I feel there is a cleaner way to handle this.

Reply:
I am sorry I did not take that as the cancelling of the polling but instead as the cancelling of the method as a whole. But I still think there is a cleaner way to rewrite the polling goroutine while keeping the mechanism above as is. I might be totally wrong though as I am only visualizing the code in my head 😄

Reply:
You're going to have to give me a picture of what you are imagining. I don't feel like I have enough hints in what you have written to give me a picture of what you imagine would work.

Additionally, I reread the context package docs and I feel the use here is in line with the guidance there.
> Package context defines the Context type, which carries deadlines, **cancellation signals**, and other request-scoped values across API boundaries and between processes.
I recall plenty of code where the code exits immediately after `<- ctx.Done` but I've seen a fair bit where calling `cancel()` changes the path of the code to a sort of cleanup operation, similar to what is written here.

To be super clear here, It's not that I think I'm right, it's that I don't really have a clear picture of what you are imagining. I am trying to explore the value of this solution to give you an opportunity to share a clearer picture of the other solution you are suggesting.

With that all said, I think I can see a way to refactor this code to make the context canceling of `pollRead()` more clear. I'll leave this unresolved until you review that.

Reply:
Okay @NajiObeid how does the new use of `context` and `cancel()` look to you?

Reply:
I think there is still room for improvement, but I'd encourage us to move along because we'll have a better chance to explore the concurrency ramifications when considering how the context here interacts with the context passed to the `Query` service method at the grpc level.

Reply:
Ok, I'll keep that in mind and we can move on.

Reply:
That's fair. I am sorry if that got misinterpreted but I never expect anyone to do some mental gymnastics to try and guess what code to write. It's just that I saw an opportunity to improve the code from a simplicity and readability perspective and wanted to see how you'd approach it.

> Additionally, I reread the context package docs and I feel the use here is in line with the guidance there.

Yes, I am sorry if that also got misunderstood in my previous comment. I am good with this approach if it makes, it's just that my initial assumption leaned toward stopping instead of cleaning.

Reply:
Ahhh, I think I understand your direction now. I'll keep all this in mind as we move on. 

Also, I definitely would agree with you, I'm not in love with pollRead/finishRead. I don't know if I see right off the bat how I want to improve it, but I'll see if something comes up.

As a final note, since `rpc Query()` is a server streaming request there isn't a context available. When we get there I'll add a comment to make sure I haven't missed something.